Sources/MongoSwift/BSON/Document.swift

```diff
@@ -157,17 +164,34 @@ extension Document {
 
                 // otherwise, we just create a new document and replace this key
             } else {
-                // TODO: SWIFT-224: use va_list variant of bson_copy_to_excluding to improve performance
-                var newSelf = Document()
+                guard let iter = DocumentIterator(forDocument: self) else {
+                    throw Document.BSONBufferTooSmallError
+                }
+
+                var keysBefore: [String] = []
+                var keysAfter: [String] = []
                 var seen = false
-                try self.forEach { pair in
-                    if !seen && pair.key == key {
+
+                while let next = iter.next() {
+                    if next.key == key {
                         seen = true
-                        try newSelf.setValue(for: pair.key, to: newValue)
+                        keysBefore.append(next.key)
+                        keysAfter.append(next.key)
                     } else {
-                        try newSelf.setValue(for: pair.key, to: pair.value)
+                        seen ? keysAfter.append(next.key) : keysBefore.append(next.key)
                     }
                 }
+
+                // To preserve order, we construct a Document excluding keys after the key, set the new value for the
+                // key, and then append all keys after the key.
+                guard let bson = bson_new() else {
+                    throw Document.InvalidBSONError
+                }
+
+                try DocumentIterator.copyElements(from: self._bson, to: bson, excluding: keysAfter)
+                var newSelf = Document(stealing: bson)
+                try newSelf.setValue(for: key, to: newValue)
```

Comment:
we can pass in `checkForKey: false` here to skip the key checking branch, since we know we excluded it in the previous statement.

Reply:
Done.